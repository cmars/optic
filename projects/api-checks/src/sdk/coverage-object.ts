type IPath = (string | number)[];

type HasCoverage = { _coverage: IPath[] };
export function wrapObjectInCoverageProxy<A>(
  obj: A,
  base: IPath = [],
  paths: Set<IPath> = new Set()
): A & HasCoverage {
  const handler = {
    get(target: any, prop: string | symbol) {
      if (prop === "_coverage") {
        return Array.from(paths).sort();
      }

      const child = Reflect.get(target, prop);
      const thisPath = [...base, prop.toString()];

      if (child) {
        if (Array.isArray(target) && isNaN(Number(prop.toString()))) {
          // skip
        } else {
          paths.add(thisPath);
        }
      }
      if (child && (isObject(child) || Array.isArray(child))) {
        return wrapObjectInCoverageProxy(child, thisPath, paths);
      } else {
        return child;
      }
    },
  };

  return new Proxy(obj, handler) as A & HasCoverage;
}

function isObject(a: any) {
  return !!a && a.constructor === Object;
}
