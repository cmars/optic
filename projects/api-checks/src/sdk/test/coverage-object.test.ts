import { wrapObjectInCoverageProxy } from "../coverage-object";

describe("object tracking its coverage", () => {
  const example = [
    {
      country: "NA",
      name: "City Name",
      lat: "44.46372",
      stats: {
        population: [
          {
            year: 2020,
            scalar: 4.3,
            unit: "million",
          },
          {
            year: 2010,
            scalar: 4.3,
            unit: "million",
          },
        ],
      },
      lng: "1.49129",
    },
  ];

  it("can return json paths", () => {
    const coverageTrackingObject = wrapObjectInCoverageProxy(example);
    coverageTrackingObject.forEach((i) => {
      const lat = i.lat;
      const population = i.stats.population;
    });

    expect(coverageTrackingObject._coverage).toMatchSnapshot();
  });
});
