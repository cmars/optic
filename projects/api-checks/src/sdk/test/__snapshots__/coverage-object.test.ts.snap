// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`object tracking its coverage can return json paths 1`] = `
Array [
  Array [
    "0",
  ],
  Array [
    "0",
    "lat",
  ],
  Array [
    "0",
    "stats",
  ],
  Array [
    "0",
    "stats",
    "population",
  ],
]
`;
